增加了校园卡余额查询即校园卡交易查询的功能

ZhuLegend opened the pull request.

Hagb commented on mycqu/card.py:
```diff
@@ -201,5 +289,37 @@ def get_fee_data(synjones_auth, room, fee_item_id):
+
+    data = {
+        'sdate': start_date.strftime('%Y-%m-%d'),
+        'edate': end_date.strftime('%Y-%m-%d'),
```
如果运行环境的时区配置不在北京时间东八区，这个这里 `strftime` 可能会得到其他时区下的时间。
可以 `from .utils.datetimes import TIMEZONE`，然后获取当前时间时用 `datetime.datetime.now(tz=TIMEZONE)`

Hagb commented on mycqu/card.py:
```diff
@@ -6,11 +6,13 @@
 import requests
 from requests import Session
 import json
+import time
```
`time` 似乎并没有被用到。

Hagb replied:
```suggestion
```

Hagb commented on mycqu/card.py:
```diff
@@ -6,11 +6,13 @@
 from html.parser import HTMLParser
 
-__all__ = ("EnergyFees",)
+__all__ = ("EnergyFees", "Bill", "Card")
```
这里最好把各个异常也加上。

Hagb commented on mycqu/card.py:
```diff
@@ -37,12 +39,12 @@ class ParseError(Exception):
 
 
-class FeeAcquisitionFailed(Exception):
+class AcquisitionFailed(Exception):
```
仔细一想这个词（`Acquisition`）好像不太常见，而且似乎这个异常和 `CQUWebsiteError` 是重合的。（这个 PR 合了之后我打算重构一下，把异常都放到一个文件里了）

Hagb replied:
如果使用 `CQUWebsiteError`，此处可以先从 `score` 导入 `CQUWebsiteError`，或者就在这个 PR 里面新建一个 `exceptions.py` 将 `CQUWebsiteError` 移进去

Hagb replied:
如果使用 `CQUWebsiteError`，此处可以先从 `score` 导入 `CQUWebsiteError`：

```suggestion
from score import CQUWebsiteError
```

或者就在这个 PR 里面新建一个 `exceptions.py` 将 `CQUWebsiteError` 移进去

Hagb commented on mycqu/card.py:
```diff
@@ -201,5 +289,37 @@ def get_fee_data(synjones_auth, room, fee_item_id):
+
+
+# 获取校园卡账号信息
+def get_card_data(session):
+    url = "http://card.cqu.edu.cn/NcAccType/GetCurrentAccountList"
+
+    res = session.post(url)
+    result = json.loads(json.loads(res.text))
+    if result['respCode'] != "0000":
+        raise AcquisitionFailed(error_msg=result['respInfo'])
+
+    return result['objs']
```
感觉按照以往的命名风格这个更像 `get_card_raw`?
```suggestion
def get_card_raw(session):
    url = "http://card.cqu.edu.cn/NcAccType/GetCurrentAccountList"

    res = session.post(url)
    result = json.loads(json.loads(res.text))
    if result['respCode'] != "0000":
        raise AcquisitionFailed(error_msg=result['respInfo'])

    return result['objs']
```

ZhuLegend replied:
对这两个函数命名的问题，因为我之前考虑只返回card对象，card中含有List[bill]，所以用统一的get_card_raw方法来获取card生成所需的所有数据，将现在这两个函数命名为get_xxx_data，如果按照我在card和bill对象提出的新方案，这两个函数就应当命名为get_xxx_raw

Hagb commented on mycqu/card.py:
```diff
@@ -201,5 +289,37 @@ def get_fee_data(synjones_auth, room, fee_item_id):
+
+
+# 获取消费账单记录
+def get_bill_data(session, account, duration):
```
同上这个似乎更像 `get_bill_raw`
```suggestion
def get_bill_raw(session, account, duration):
```

Hagb commented on mycqu/card.py:
```diff
@@ -124,13 +144,81 @@ def fetch(session: Session, isHuxi: bool, room: str) -> EnergyFees:
+    @staticmethod
+    def from_dict(data: dict[str, Any]) -> Bill:
+        """
+        从反序列化的（一个）账单 json 中获取账单信息
+
+        :param data: json 反序列化得到的字典
+        :type data: dict[str, Any]
+        :return: 账单对象
+        :rtype: Bill
+        """
+        return Bill(
+            tran_name=data['tranName'],
+            tran_date=data['tranDt'],
+            tran_place=data['mchAcctName'],
+            tran_amount=float(data['tranAmt'] / 100),
+            acc_amount=float(int(data['acctAmt']) / 100)
+        )
+
+
+@dataclass
+class Card:
+    """
+    校园卡及其账单信息
+    """
+    amount: float
+    """账户余额"""
+    bills: List[Bill]
+    """最近交易记录"""
+
+    @staticmethod
+    def fetch(session: Session, duration: int = 30) -> Card:
+        res = get_bill_raw(session, duration)
+        bills = []
+        for bill in res['bills']:
+            bills.append(Bill.from_dict(bill))
+
+        return Card(
+            amount=float(res['amount'] / 100),
+            bills=bills
+        )
```
从设计的角度来说我倾向于 `Card` 只存卡号和余额，而交易记录 `Bill`（或者说 `List[Bill]`）的获取则依赖于对应的 `Card` 对象或卡号。好处是，不需要交易记录而仅仅需要卡号时候，可以少进行一个请求。

ZhuLegend replied:
我在设计的时候有考虑让bill依赖于card对象，但是最后没这么做的原因是：虽然bill依赖于card对象可以减少对教务网的访问，但是当需要账单的时候需要分别调用不同对象的fetch方法，而且bill对象还需要card对象作为参数，这增加了理解成本，所以最后我采用了当前pr提交的方式
但是我刚刚想到了一个新的思路，或许给card对象添加一个fetchBill方法，返回一个List[Bill]，在调用用fetch后对返回的card对象调用这个方法就可以拿到这张卡的对应消费记录。或许这是一个更好的方案？

Hagb replied:
> 但是我刚刚想到了一个新的思路，或许给card对象添加一个fetchBill方法，返回一个List[Bill]，在调用用fetch后对返回的card对象调用这个方法就可以拿到这张卡的对应消费记录。或许这是一个更好的方案？

我觉得这个思路挺好（另，snake case 风格命名下应该是 `fetch_bill`）

Hagb commented on mycqu/card.py:
```diff
@@ -124,13 +144,81 @@ def fetch(session: Session, isHuxi: bool, room: str) -> EnergyFees:
+
+
+# 使用登陆统一认证的会话登陆查询网站的辅助函数
+def card_login(session):
+    res = session.get(LOGIN_URL)
+
+    parser = CardPageParser()
+    parser.feed(res.text)
+    ssoticket_id = parser.ssoticket_id
+    get_hall_ticket(session, ssoticket_id)
```
从尽量减少对学校网站请求数的角度来看的话，我觉得这个可以像 `access_mycqu` 那样子，其他的查询 API 依赖于它（即，需要开发者先调用这个登录的 api）但不显式调用它。

Hagb reviewed (commented):
这是个人对这里的 API 设计的一些看法，供参考（不一定要采用）。
主要的出发点是
- 和已有 API 风格的统一
- 尽量减小对学校网站的请求

Hagb commented on mycqu/card.py:
```diff
@@ -124,13 +144,81 @@ def fetch(session: Session, isHuxi: bool, room: str) -> EnergyFees:
+    tran_name: str
+    """交易名称"""
+    tran_date: str
+    """交易时间"""
```
这里可以使用 python 的 `datetime.date` 对象

Hagb commented on mycqu/exception.py:
```diff
@@ -0,0 +1,90 @@
+class NetworkError(Exception):
+    """
+    当访问相关网页时statue code不为200时抛出
+    """
```
我觉得这个其实不能叫“网络错误”，因为这个不是由于网络原因引起的（网络原因是诸如“连接超时”这类的错误）。
我个人倾向于把使用这个异常类的情况归到 `CQUWebsiteError`，或者作为 `CQUWebsiteError` 的子类（名字大概要改一改）

Hagb commented on mycqu/card.py:
```diff
@@ -190,46 +176,44 @@ class Card:
+            amount=float(card_info['acctAmt'] / 100)
         )
 
+    def fetch_bill(self, session: Session) -> List[Bill]:
```
抱歉，先前没意识到 `bill` 是个可数名词…
```suggestion
    def fetch_bills(self, session: Session) -> List[Bill]:
```

Hagb commented on mycqu/card.py:
```diff
@@ -178,7 +164,7 @@ def from_dict(data: dict[str, Any]) -> Bill:
         return Bill(
             tran_name=data['tranName'],
-            tran_date=data['tranDt'],
+            tran_date=datetime.datetime.strptime(data['tranDt'], "%Y-%m-%d %H:%M:%S"),
```
直接 `strptime` 得到的时间也是会按照运行环境的时区，可以用 `replace` 将北京时间应用上去：
```suggestion
            tran_date=datetime.datetime.strptime(data['tranDt'], "%Y-%m-%d %H:%M:%S").replace(tzinfo=TIMEZONE),
```
不过全堆在这行似乎长了点，也可以换种写法…

Hagb commented on mycqu/mycqu.py:
```diff
@@ -51,3 +50,46 @@ def access_mycqu(session: Session, add_to_header: bool = True) -> Dict[str, str]
+    """用登陆了统一身份认证的会话在 card.cqu.edu.cn 进行认证
+
+    :param session: 登陆了统一身份认证的会话
+    :type session: Session
+    """
+    res = session.get(LOGIN_URL)
+
+    parser = CardPageParser()
+    parser.feed(res.text)
+    ssoticket_id = parser.ssoticket_id
+    get_hall_ticket(session, ssoticket_id)
+
+
+class CardPageParser(HTMLParser):
+    def __init__(self):
+        super().__init__()
+        self._starttag: bool = False
+        self.ssoticket_id: str = ""
+
+    def handle_starttag(self, tag, attrs):
+        if not self._starttag and tag == 'input' and ('name', 'ssoticketid') in attrs:
+            self._starttag = True
+            for key, val in attrs:
+                if key == "value":
+                    self.ssoticket_id = val
+                    break
+
+
+# 获取hallticket
+def get_hall_ticket(session, ssoticket_id):
+    url = 'http://card.cqu.edu.cn/cassyno/index'
+    data = {
+        'errorcode': '1',
+        'continueurl': 'http://card.cqu.edu.cn/cassyno/index',
+        'ssoticketid': ssoticket_id,
+    }
+    r = session.post(url, data=data)
+    if r.status_code != 200:
+        raise NetworkError()
+    return session
```
我个人认为 `card_login`（及其相关内容）放这里不一定合适，因为 <http://card.cqu.edu.cn> 和 <https://my.cqu.edu.cn> 并没有从属或者依赖的关系。

刚刚注意到，<http://card.cqu.edu.cn> 还有另一套它自己的登录方式（非统一身份认证登录），如果以后要加入这个登录方式的支持的话，现在 `card_login` 这个函数的名字是不是还需要再斟酌一下比如 `access_card`。（这段供参考）
BTW: 我试了试这个登录方式，和统一身份认证登录比起来似乎没什么优势，尤其是登录的时候还遇到了一个时长 21 秒的请求…

ZhuLegend replied:
我把它移到这里的原因是他们都需要一个登陆了统一认证的会话作为参数（这应该手机依赖于my.cqu.edu.cn了的。但确实和mycqu.py这个文件名不太匹配，或许给mycqu这个文件夹换个名字更好？）

我认为它原生到登陆方式放到card.py里面才差不多（但是原生登陆可能碰到超长延迟的话可能就不太有写的动力了……）

Hagb replied:
我解释一下我那个意思：
my.cqu.edu.cn 的登录依赖于统一身份认证 authserver.cqu.edu.cn，后者不依赖于前者。
card.cqu.edu.cn 同理（不过 card.cqu.edu.cn 又有另外一种登录方式，实不实现另说）。

然后 my.cqu.edu.cn 和 card.cqu.edu.cn 是并列的关系。

`access_mycqu` 放 mycqu.py 的原因是它是给 my.cqu.edu.cn 用的，同时 authserver.cqu.edu.cn 并不依赖于 `access_mycqu`（换句话说用统一身份认证并不一定会用 my.cqu.edu.cn）

（同样的逻辑应用到 card 上的话，这里的 `card_login` 可以是放在 `card.py` 的 `access_card`）

ZhuLegend replied:
啊，我authserver和my的关系记错了……就我的意思是，card_login和access_mycqu都是需要登陆统一身份认证之后才调用，所以应该放在一起更好？（所以我的建议就是直接把mycqu.py的名字改了）
当然，直接改成access_card也行，但是就是会显得mycqu.py比card.py文件单薄很多……（毕竟mycqu里面就一个登陆mycqu的功能，card里面已经有挺多的函数了）

Hagb replied:
> 啊，我authserver和my的关系记错了……就我的意思是，card_login和access_mycqu都是需要登陆统一身份认证之后才调用，所以应该放在一起更好？（所以我的建议就是直接把mycqu.py的名字改了）

我觉得这个没问题，这样子差不多是按功能来分

> 当然，直接改成access_card也行，但是就是会显得mycqu.py比card.py文件单薄很多……（毕竟mycqu里面就一个登陆mycqu的功能，card里面已经有挺多的函数了）

还有一种方式是，关于 `mycqu` 的功能建一个文件夹 `mycqu`（或者 `my`），`access_mycqu` 可以放在这个文件夹下的 `__init__.py`，其他 my.cqu.edu.cn 的功能则放在其中不同 `py` 文件里（向下兼容可以通过模块根目录下 `__init__.py` 的 `import` 来做）。这样子好处是如果之后如果要包含学校多个系统的话，可以直接按域名来划分结构（比如 card.cqu.edu.cn、my.cqu.edu.cn），然后要导入某个系统的多个功能时也比较方便。

Hagb commented:
LGTM，非常感谢！请问可以合并了不？

ZhuLegend committed:
Add support for card balance and transaction query

ZhuLegend committed:
refactor: create exceptions.py and refactor card.py

- All exceptions are moved to exceptions.py
- Use the same coding style in card.py
